### ISSUE TITLE
[QUERY] Azure.Messaging.ServiceBus -  How does an exception occurring in GetTokenAsync handler get handled by the library

### ISSUE THREAD

**Issue by chris-skuvault** [Azure/azure-sdk-for-net#18942]
**Query/Question**
How can we help?

We use the TokenCredential overload when creating a ServiceBusClient to manage auth for use with service bus. We recently started to see some intermittent issues when calling _confidentialClient.AcquireTokenForClient to retrieve the access token needed for ServiceBus. Sometime this call starts to return 503's which we have a policy in place to retry but occasionally it goes over the retry limit we have set in place and we let the exception bubble after that into Azure.Messaging.ServiceBus library. 

Looking at Azure.Messaging.ServiceBus and how it handles this I've traced it down to https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConnectionScope.cs#L915 throwing in this method. This method catches the exception and sets the refreshTimer's dueTime to Timeout.Infinte. There is a comment present that indicates the operation would get retried but I'm having trouble following how this works in regards to an exception being thrown at this point.

![image](https://user-images.githubusercontent.com/40215033/108877986-c5ab1080-75cd-11eb-86a9-bcee873fb282.png)

What is the correct way to handle this scenario and does anything else need to be done to ensure proper retries. Should we the caller keep retrying on 503's indefinitely and not let the exception bubble up at all? Or does things get handled properly already and nothing needs to be done?

Here is stack from when issue occured
```csharp
MSAL.NetCore.4.19.0.0.MsalServiceException: 
	ErrorCode: service_not_available
Microsoft.Identity.Client.MsalServiceException: Service is unavailable to process the request
   at Microsoft.Identity.Client.Http.HttpManager.ExecuteWithRetryAsync(Uri endpoint, IDictionary`2 headers, HttpContent body, HttpMethod method, ICoreLogger logger, Boolean doNotThrow, Boolean retry)
   at Microsoft.Identity.Client.Http.HttpManager.ExecuteWithRetryAsync(Uri endpoint, IDictionary`2 headers, HttpContent body, HttpMethod method, ICoreLogger logger, Boolean doNotThrow, Boolean retry)
   at Microsoft.Identity.Client.Http.HttpManager.SendPostAsync(Uri endpoint, IDictionary`2 headers, HttpContent body, ICoreLogger logger)
   at Microsoft.Identity.Client.Http.HttpManager.SendPostAsync(Uri endpoint, IDictionary`2 headers, IDictionary`2 bodyParameters, ICoreLogger logger)
   at Microsoft.Identity.Client.OAuth2.OAuth2Client.ExecuteRequestAsync[T](Uri endPoint, HttpMethod method, RequestContext requestContext, Boolean expectErrorsOn200OK, Boolean addCommonHeaders)
   at Microsoft.Identity.Client.OAuth2.OAuth2Client.GetTokenAsync(Uri endPoint, RequestContext requestContext, Boolean addCommonHeaders)
   at Microsoft.Identity.Client.OAuth2.TokenClient.SendHttpAndClearTelemetryAsync(String tokenEndpoint)
   at Microsoft.Identity.Client.OAuth2.TokenClient.SendHttpAndClearTelemetryAsync(String tokenEndpoint)
   at Microsoft.Identity.Client.OAuth2.TokenClient.SendTokenRequestAsync(IDictionary`2 additionalBodyParameters, String scopeOverride, String tokenEndpointOverride, CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.SendTokenRequestAsync(String tokenEndpoint, IDictionary`2 additionalBodyParameters, CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.ClientCredentialRequest.FetchNewAccessTokenAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.ClientCredentialRequest.ExecuteAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.RunAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.ApiConfig.Executors.ConfidentialClientExecutor.ExecuteAsync(AcquireTokenCommonParameters commonParameters, AcquireTokenForClientParameters clientParameters, CancellationToken cancellationToken)
   at SkuVault.ServiceBus.Management.ServiceBusTokenService.<>c__DisplayClass11_0.<<AcquireTokenAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Polly.Retry.AsyncRetryEngine.ImplementationAsync[TResult](Func`3 action, Context context, CancellationToken cancellationToken, ExceptionPredicates shouldRetryExceptionPredicates, ResultPredicates`1 shouldRetryResultPredicates, Func`5 onRetryAsync, Int32 permittedRetryCount, IEnumerable`1 sleepDurationsEnumerable, Func`4 sleepDurationProvider, Boolean continueOnCapturedContext)
   at Polly.AsyncPolicy`1.ExecuteAsync(Func`3 action, Context context, CancellationToken cancellationToken, Boolean continueOnCapturedContext)
   at SkuVault.ServiceBus.Management.ServiceBusTokenService.AcquireTokenAsync(CancellationToken token, List`1 scopes)
   at SkuVault.ServiceBus.Management.ServiceBusTokenService.GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
   at Azure.Messaging.ServiceBus.Amqp.CbsTokenProvider.GetTokenAsync(Uri namespaceAddress, String appliesTo, String[] requiredClaims)
   at Microsoft.Azure.Amqp.TaskHelpers.EndAsyncResult(IAsyncResult asyncResult)
   at Microsoft.Azure.Amqp.IteratorAsyncResult`1.StepCallback(IAsyncResult result)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Amqp.AsyncResult.End[TAsyncResult](IAsyncResult result)
   at Microsoft.Azure.Amqp.AmqpCbsLink.<>c__DisplayClass4_0.<SendTokenAsync>b__1(IAsyncResult a)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
   at Azure.Messaging.ServiceBus.Amqp.AmqpConnectionScope.RequestAuthorizationUsingCbsAsync(AmqpConnection connection, CbsTokenProvider tokenProvider, Uri endpoint, String[] audience, String[] requiredClaims, TimeSpan timeout)
   at Azure.Messaging.ServiceBus.Amqp.AmqpConnectionScope.<>c__DisplayClass64_0.<<CreateAuthorizationRefreshHandler>b__0>d.MoveNext()
	StatusCode: 503 
	ResponseBody:  
	Headers: Cache-Control: private
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Content-Type-Options: nosniff
client-request-id: 44390ff3-e528-4a60-9630-8338a2b00333
x-ms-request-id: e7f01568-0a41-4e9d-a5fa-2da1f017fd00
x-ms-ests-server: 2.1.11496.7 - NCUS ProdSlices
Set-Cookie: x-ms-gateway-slice=estsfd; path=/; secure; httponly, stsservicecookie=estsfd; path=/; secure; samesite=none; httponly
Date: Tue, 23 Feb 2021 15:03:12 GMT
```

**Environment:**
 - Azure.Messaging.ServiceBus 7.0.0
 - Windows .net core 3.1

**Comment by jsquire** [comment]
Hi @chris-skuvault.  Thank you for your feedback and we regret that you're experiencing difficulties.  I'll investigate and follow-up tomorrow with details.

**Comment by jsquire** [comment]
Your approach for retrying token acquisition and bubbling the exception if unsuccessful when retries are exhausted is the approach that we'd recommend for custom token providers.

In the failure flow when a token is unavailable, the Service Bus client will stop attempting to refresh authorization on the active link.  The link will be considered invalid by the Service Bus service and force-closed.  _(if we kept attempting renewal, there's a good chance that authorization would have expired and the link would have already been rejected by the service,  resulting in the same flow)_

When an operation is attempted on the client, it attempts to retrieve an AMQP link, following a get-or-create pattern.   In the case that there is no active link  (or the link is faulted), a new link will be opened which includes sending a fresh token for authorization and establishing a new timer to refresh auth when that token is near expiration.  In the event that a token is unavailable when the client attempts to create a new link, the operation will fail and an `UnauthorizedAccessException` will be thrown.

There is a gap in the current logic in that we do not consider the `UnauthorizedAccessException` retriable - which leaves the client susceptible to a race condition in the event that authorization expires for a link while the client is in the middle of performing an operation.   In that scenario, the client would think it had a healthy active AMQP link to use and the service would reject the operation as unauthorized.   That also degrades the experience in the flow for opening a link.

We're planning on addressing the oversight on recognizing the `UnauthorizedAccessException` during the current milestone, and I expect that the fix will be available in our next release.   This is being tracked [here](https://github.com/Azure/azure-sdk-for-net/issues/16419), though I'll likely address it as a stand-alone change at some point this week.

**Comment by chris-skuvault** [comment]
@jsquire Thanks for the insight. Thank helps me understand a lot. 

As far as the gap you mentioned with the `UnauthorizedAccessException` we do occasionally run into this race condition I believe as we get intermittent errors where we have to recreate the session receiver to continue processing the session. 

```csharp
Azure.Messaging.ServiceBus.ServiceBusException: The session lock was lost. Request a new session receiver. (SessionLockLost)
 ---> Azure.Messaging.ServiceBus.ServiceBusException: The link 'amqps://prod-servicebus-5.servicebus.windows.net/-907072a9;2:8434:8449:source(address:/priority-6,filter:[com.microsoft:session-filter:352673])' is force detached. Code: RenewToken. Details: Unauthorized access. 'Listen' claim(s) are required to perform this operation. Resource: 'sb://prod-servicebus-5.servicebus.windows.net/priority-6'.. TrackingId:e7156aa3b8b74bd99948c3af8f2c7650_G3, SystemTracker:gateway7, Timestamp:2021-02-20T05:30:05 (GeneralError)
   --- End of inner exception stack trace ---
   at Azure.Messaging.ServiceBus.Amqp.AmqpReceiver.ThrowIfSessionLockLost()
   at Azure.Messaging.ServiceBus.Amqp.AmqpReceiver.ExecuteRequest(TimeSpan timeout, AmqpRequestMessage amqpRequestMessage)
   at Azure.Messaging.ServiceBus.Amqp.AmqpReceiver.RenewSessionLockInternal(TimeSpan timeout)
   at Azure.Messaging.ServiceBus.Amqp.AmqpReceiver.<>c__DisplayClass56_0.<<RenewSessionLockAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Azure.Messaging.ServiceBus.ServiceBusRetryPolicy.RunOperation(Func`2 operation, TransportConnectionScope scope, CancellationToken cancellationToken)
   at Azure.Messaging.ServiceBus.ServiceBusRetryPolicy.RunOperation(Func`2 operation, TransportConnectionScope scope, CancellationToken cancellationToken)
   at Azure.Messaging.ServiceBus.Amqp.AmqpReceiver.RenewSessionLockAsync(CancellationToken cancellationToken)
   at Azure.Messaging.ServiceBus.ServiceBusSessionReceiver.RenewSessionLockAsync(CancellationToken cancellationToken)
   at SkuVault.ServiceBus.Receiver.SessionPriorityReceiver.RenewSessionLockAsync()
```

Would this be an indication of this issue?

**Comment by jsquire** [comment]
Hi @chris-skuvault.   That looks to be the race condition that I was describing.   The link was open when the operation was attempted but the token had expired.   I just merged in a fix that will help guard against that race in most conditions.  (#19041)

Unfortunately, there's a catch.  I'm seeing that you're in a scenario where sessions are in use.  The semantics around sessions are different in that reconnection is not allowed, so this won't help with what you're seeing in the above snippet.    @JoshLove-msft has a much deeper knowledge of the quirks around sessions.    Josh:  do you have any suggestions or thoughts in this case?

**Comment by JoshLove-msft** [comment]
Yes, with sessions, we don't reconnect the link. This is so that we can provide more deterministic behavior in the rest of the operations in a session receiver. Imagine that we did attempt to re-connect - you would need to handle the scenario where a SessionCannotBeLocked error is thrown, (which might be confusing given that you just had the session lock) in addition to a ServiceTimeout error (since there may be no more messages in that session) when doing ANY operation with the SessionReceiver. Rather than spread those exceptions out to all operations in the session receiver, we decided to have a central place where these are expected outcomes - when first accepting a session.
